transport/sqs/src/main/kotlin/SqsMessageReceiverFactory.kt

```diff
@@ -0,0 +1,125 @@
+import org.eclipse.apoapsis.ortserver.transport.MessageHeader
+import org.eclipse.apoapsis.ortserver.transport.MessageReceiverFactory
+import org.eclipse.apoapsis.ortserver.transport.RUN_ID_PROPERTY
+import org.eclipse.apoapsis.ortserver.transport.TOKEN_PROPERTY
+import org.eclipse.apoapsis.ortserver.transport.TRACE_PROPERTY
+import org.eclipse.apoapsis.ortserver.transport.json.JsonSerializer
+
+import org.slf4j.LoggerFactory
+
+private val logger = LoggerFactory.getLogger(SqsMessageReceiverFactory::class.java)
+
+internal val messageAttributeNames = listOf(TOKEN_PROPERTY, TRACE_PROPERTY, RUN_ID_PROPERTY)
+
+/**
+ * A [MessageReceiverFactory] implementation for AWS SQS.
+ * Also see https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/kotlin_sqs_code_examples.html.
+ */
+class SqsMessageReceiverFactory : MessageReceiverFactory {
+    override val name: String = SqsConfig.TRANSPORT_NAME
+
+    override fun <T : Any> createReceiver(
+        from: Endpoint<T>,
+        configManager: ConfigManager,
+        handler: EndpointHandler<T>
+    ) = runBlocking {
+        logger.info("Creating SQS receiver for endpoint '${from.configPrefix}'.")
+
+        val serializer = JsonSerializer.forClass(from.messageClass)
+        val config = SqsConfig.create(configManager)
+        val client = createSqsClient(config)
+
+        val queueResponse = client.getQueueUrl(GetQueueUrlRequest { queueName = config.queueName })
+
+        val queueUrl = checkNotNull(queueResponse.queueUrl) {
+            "The SQS queue URL for queue '$name' must not be null."
+        }
+
+        val receiveRequest = ReceiveMessageRequest {
+            this.queueUrl = queueUrl
+
+            // Opt-in to the message attributes to receive.
+            messageAttributeNames = org.eclipse.apoapsis.ortserver.transport.sqs.messageAttributeNames
+
+            // Enable "long polling" to eliminate empty responses, see
+            // https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/working-with-messages.html#setting-up-long-polling
+            waitTimeSeconds = 1
+        }
+
+        while (isActive) {
+            val receiveResponse = client.receiveMessage(receiveRequest)
+
+            receiveResponse.messages?.forEach { message ->
+                val attrs = checkNotNull(message.messageAttributes) {
+                    "The message attributes must not be null."
+                }
+
+                val body = checkNotNull(message.body) {
+                    "The message body must not be null."
+                }
+
+                runCatching {
+                    handler(Message(attrs.toMessageHeader(), serializer.fromJson(body)))
+
+                    val deleteRequest = DeleteMessageRequest {
+                        this.queueUrl = queueUrl
+                        receiptHandle = message.receiptHandle
+                    }
+
+                    client.deleteMessage(deleteRequest)
```

Comment:
I saw that there is a [`visibilityTimeout`](https://sdk.amazonaws.com/kotlin/api/latest/sqs/aws.sdk.kotlin.services.sqs.model/-receive-message-request/index.html) option to configure how long a message is invisible to subsequent retrieve requests, but at least the docs for the Kotlin SDK do not explain the behavior when this is set to `null` and what is the default. Could it happen that the message is received by another receiver if the `handler` call takes very long?

Reply:
I have the same question. Since message processing can take really long if a worker gets executed, can it have a negative impact if the message is deleted only afterward? Should deletion be done before calling the handler and how does this play together with error handling and message re-delivery?

Reply:
Counter-question: What is ORT server's expectation here? Should every message be only retceived once, by exactly one receiver?

Reply:
Here's more documentation about the setting: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html

Reply:
> Counter-question: What is ORT server's expectation here? Should every message be only retceived once, by exactly one receiver?

In general, yes. If a receiver fails while processing a message it might make sense that another receiver can pick it up later on, but this whole scenario is a bit unclear to me, and without looking at the code I don't know if other transport implementations acknowledge messages immediately or only after handling them.

Reply:
The RabbitMQ docs linked above state that a message should be acknowledged after processing it:

> Consumer acknowledgements are not aware of publishers: their goal is to confirm to a RabbitMQ node that a given delivery was successfully received and processed successfully, so the delivered message can be marked for future deletion.

However, our RabbitMQ implementation uses autoAck which acknowledges the message as soon as it was received, so it would be consistent to do the same in other transports.

Reply:
> So I'll add code to delete the message after retrieval.

I totally forgot that this [is already in place](https://github.com/eclipse-apoapsis/ort-server/pull/337/files#diff-b5c5be3aad7e14978470edd9afcb0c5d7b9063767e7820cdf2b35e4d656001ebR100-R105) 😸

Reply:
But then we are back to my question: Should the message be deleted before or after calling the handler? As it is currently, the message is not deleted if the handler crashes. Will it then be picked up again, and - if the problem persists - can an endless loop be entered?

Reply:
Quoting @mnonnenmacher:

> If a receiver fails while processing a message it might make sense that another receiver can pick it up later on

So I assumed the current behavior to delete the message only if the handler does not crash is the desired one.

> Will it then be picked up again, and - if the problem persists - can an endless loop be entered?

I believe we'd eventually have some repeated handling, but not real endless loop, as at some point the aforementioned [visibility timeout](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html) will make the message disappear. We could even [lower the visibility timeout](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html) after a message has been received but before it gets handled / deleted to reduce the number of potential repeated handlings if the handler crashes.

Just tell me how it should be implemented, and I'll do so.

Reply:
> at some point the aforementioned [visibility timeout](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html) will make the message disappear.

Scratch that, I was misreading how it works: The message will just not get delivered during that time, but will never be automatically deleted from the queue.

Reply:
Reading [more docs](https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/SQS/MSQSDeleteMessageDeleteMessageRequest.html): 

> Amazon SQS automatically deletes messages left in a queue longer than the retention period configured for the queue.

So should we simply rely on the queue's retention period here, or should we always delete the message before it's being handled?

Reply:
I'm deleting the message now before it's being passed to the handler.